### Задание 1

In [1]:
import re

In [2]:
with open('zhivago.txt', encoding='utf-8') as f:
    text = f.read()

In [3]:
def remove_trash(text):
    no_tags_text = re.sub(r'<[^>]+>', ' ', text)
    no_space_sequences_text = re.sub(r'[\s\\xa0]+', ' ', no_tags_text).strip()
    text_till_last_sentence = f"{no_space_sequences_text.rsplit('.', 1)[0]}."
    return text_till_last_sentence

In [4]:
text = remove_trash(text)

In [5]:
text[:1500]

'Борис Леонидович Пастернак Доктор Живаго «Доктор Живаго» - итоговое произведение Бориса Пастернака, книга всей его жизни. Этот роман принес его автору мировую известность и Нобелевскую премию, присуждение которой обернулось для поэта оголтелой политической травлей, обвинениями в «измене Родине» и в результате стоило ему жизни. «Доктор Живаго» - роман, сама ткань которого убедительнее свидетельствует о чуде, чем все размышления доктора и обобщения автора. Человек, который так пишет, бесконечно много пережил и передумал, и главные его чувства на свете - восхищенное умиление и слезное сострадание; конечно, есть в его мире место и презрению, и холодному отстранению - но не в них суть. Роман Пастернака - оплакивание прежних заблуждений и их жертв; те, кто не разделяет молитвенного восторга перед миром, достойны прежде всего жалости. Перечитывать «Доктора Живаго» стоит именно тогда, когда кажется, что жить не стоит. Тогда десять строк из этого романа могут сделать то же, что делает любовь в

### Задание 2

In [6]:
from collections import Counter
from string import punctuation
from rusenttokenize import ru_sent_tokenize
from razdel import tokenize as razdel_tokenize

In [7]:
sentences = ru_sent_tokenize(text)

In [212]:
sentences[200:210]

['Мужику дай волю, так ведь у нас друг дружку передавят, истинный Господь.',
 'Ай заснули?',
 'Это была вторая поездка дяди и племянника в Дуплянку.',
 'Юра думал, что он запомнил дорогу, и всякий раз, как поля разбегались вширь и их тоненькой каемкой охватывали спереди и сзади леса, Юре казалось, что он узнает то место, с которого дорога должна повернуть вправо, а с поворота показаться и через минуту скрыться десятиверстная кологривовская панорама с блещущей вдали рекой и пробегающей за ней железной дорогой.',
 'Но он все обманывался.',
 'Поля сменялись полями.',
 'Их вновь и вновь охватывали леса.',
 'Смена этих просторов настраивала на широкий лад.',
 'Хотелось мечтать и думать о будущем.',
 'Ни одна из книг, прославивших впоследствии Николая Николаевича, не была еще написана.']

In [9]:
text = text.lower()

In [10]:
punctuation = f'{punctuation}«»–...'

In [11]:
tokens = [token.text.strip(punctuation) for token in razdel_tokenize(text) 
          if token.text not in punctuation]
tokens = [token for token in tokens if token]

In [13]:
tokens[:10]

['борис',
 'леонидович',
 'пастернак',
 'доктор',
 'живаго',
 'доктор',
 'живаго',
 'итоговое',
 'произведение',
 'бориса']

In [14]:
# есть ли в тексте повторяющиеся корректные предложения? если да то какие?

In [214]:
def sorted_by_freq(lst):
    counted_lst = dict(Counter(lst)).items()
    sorted_by_freq_counted_lst = sorted(counted_lst, key=lambda x: x[1], reverse=True)
    return sorted_by_freq_counted_lst

In [215]:
[sent for sent in sorted_by_freq(sentences) if sent[1] > 1]

[('– Да.', 10),
 ('Да.', 10),
 ('Не правда ли?', 9),
 ('А?', 7),
 ('– Нет.', 7),
 ('– Хорошо.', 6),
 ('– Знаю.', 4),
 ('– Разумеется.', 4),
 ('Сеялки.', 4),
 ('Молотилки».', 4),
 ('Ха-ха-ха!', 3),
 ('– А как же.', 3),
 ('Конечно.', 3),
 ('Где он?', 3),
 ('Погоди.', 3),
 ('Маркел!', 3),
 ('Но дело не в этом.', 3),
 ('Слушай.', 3),
 ('Пойдем.', 3),
 ('Дядя Воронюк!', 3),
 ('– Еще бы.', 3),
 ('– Я знаю.', 3),
 ('Ну да!', 3),
 ('Свеча горела на столе, Свеча горела.', 3),
 ('Единственно живое и яркое в вас – это то, что вы жили в одно время со мной и меня знали».',
  2),
 ('Весной в несколько дней лес преображается, подымается до облаков, в его покрытых листьями дебрях можно затеряться, спрятаться.',
  2),
 ('Это превращение достигается движением, по стремительности превосходящим движения животных, потому что животное не растет так быстро, как растение, и которого никогда нельзя подсмотреть.',
  2),
 ('Лес не передвигается, мы не можем его накрыть, подстеречь за переменою места.',
  2),
 ('

In [17]:
# какой самый частотный токен в тексте длиннее 6 символов?

In [216]:
[word for word in sorted_by_freq(tokens) if len(word[0]) > 6][0]

('андреевич', 289)

### Задание 3

In [54]:
from nltk.stem.snowball import SnowballStemmer

In [55]:
stemmer = SnowballStemmer('russian')

In [57]:
stems = [(token, stemmer.stem(token)) for token in tokens]

In [58]:
# два разных слова ошибочно свелись к одинаковой основе

In [130]:
same_stems_for_diff_tokens = [(stem, another_stem) for stem in stems for another_stem in list(set(stems))
                              if stem[0] != another_stem[0] and stem[1] == another_stem[1]]

In [217]:
[pair for pair in sorted_by_freq(same_stems_for_diff_tokens)][:40]

[((('не', 'не'), ('нее', 'не')), 2933),
 ((('не', 'не'), ('нею', 'не')), 2933),
 ((('не', 'не'), ('ней', 'не')), 2933),
 ((('он', 'он'), ('оно', 'он')), 1577),
 ((('он', 'он'), ('они', 'он')), 1577),
 ((('он', 'он'), ('она', 'он')), 1577),
 ((('он', 'он'), ('оной', 'он')), 1577),
 ((('как', 'как'), ('каки', 'как')), 1288),
 ((('как', 'как'), ('каков', 'как')), 1288),
 ((('как', 'как'), ('какому', 'как')), 1288),
 ((('как', 'как'), ('каком', 'как')), 1288),
 ((('как', 'как'), ('каким', 'как')), 1288),
 ((('как', 'как'), ('какою', 'как')), 1288),
 ((('как', 'как'), ('какого', 'как')), 1288),
 ((('как', 'как'), ('какая', 'как')), 1288),
 ((('как', 'как'), ('какую', 'как')), 1288),
 ((('как', 'как'), ('какой', 'как')), 1288),
 ((('как', 'как'), ('кака', 'как')), 1288),
 ((('как', 'как'), ('каких', 'как')), 1288),
 ((('как', 'как'), ('какими', 'как')), 1288),
 ((('как', 'как'), ('какие', 'как')), 1288),
 ((('как', 'как'), ('какое', 'как')), 1288),
 ((('это', 'эт'), ('этих', 'эт')), 1001),
 

Из частотных пар, подходящих под условие, ошибка в стемминге есть в следующих:  
_((('не', 'не'), ('нее', 'не')), 2933),   
 ((('не', 'не'), ('нею', 'не')), 2933),   
 ((('не', 'не'), ('ней', 'не')), 2933),   
 ((('а', 'а'), ('ай', 'а')), 990),  
 ((('из', 'из'), ('изо', 'из')), 796)_

В первых трех из них у второй пары внутри каждой пары основа должна быть нулевой, а в четвертой и пятой парах у вторых пар основа должна быть одинакова с токеном.

In [136]:
# слово не изменилось после стемминга (слово должно быть русским и длиннее 4 символов)

In [137]:
unchanged_tokens = [stem for stem in stems if stem[0] == stem[1] and 
                    len(stem[0]) > 4 and re.search(r'[а-яё]', stem[0])]

In [218]:
[token for token in sorted_by_freq(unchanged_tokens)][:15]

[(('андреевич', 'андреевич'), 289),
 (('доктор', 'доктор'), 226),
 (('перед', 'перед'), 172),
 (('может', 'может'), 162),
 (('через', 'через'), 144),
 (('будет', 'будет'), 141),
 (('сейчас', 'сейчас'), 127),
 (('человек', 'человек'), 112),
 (('вдруг', 'вдруг'), 109),
 (('поезд', 'поезд'), 65),
 (('видел', 'видел'), 53),
 (('назад', 'назад'), 49),
 (('гордон', 'гордон'), 47),
 (('город', 'город'), 45),
 (('хотел', 'хотел'), 45)]

Из частотных пар, подходящих под условие, ошибка в стемминге есть у слов _может, будет, человек, видел, хотел_. Правильными основами были бы соответственно _мо, б,_ нулевая, _ви, хо_.

### Задание 4

In [140]:
from nltk.corpus import stopwords

In [141]:
stops = stopwords.words('russian')

In [143]:
print(stops)

['и', 'в', 'во', 'не', 'что', 'он', 'на', 'я', 'с', 'со', 'как', 'а', 'то', 'все', 'она', 'так', 'его', 'но', 'да', 'ты', 'к', 'у', 'же', 'вы', 'за', 'бы', 'по', 'только', 'ее', 'мне', 'было', 'вот', 'от', 'меня', 'еще', 'нет', 'о', 'из', 'ему', 'теперь', 'когда', 'даже', 'ну', 'вдруг', 'ли', 'если', 'уже', 'или', 'ни', 'быть', 'был', 'него', 'до', 'вас', 'нибудь', 'опять', 'уж', 'вам', 'ведь', 'там', 'потом', 'себя', 'ничего', 'ей', 'может', 'они', 'тут', 'где', 'есть', 'надо', 'ней', 'для', 'мы', 'тебя', 'их', 'чем', 'была', 'сам', 'чтоб', 'без', 'будто', 'чего', 'раз', 'тоже', 'себе', 'под', 'будет', 'ж', 'тогда', 'кто', 'этот', 'того', 'потому', 'этого', 'какой', 'совсем', 'ним', 'здесь', 'этом', 'один', 'почти', 'мой', 'тем', 'чтобы', 'нее', 'сейчас', 'были', 'куда', 'зачем', 'всех', 'никогда', 'можно', 'при', 'наконец', 'два', 'об', 'другой', 'хоть', 'после', 'над', 'больше', 'тот', 'через', 'эти', 'нас', 'про', 'всего', 'них', 'какая', 'много', 'разве', 'три', 'эту', 'моя', 'впр

Посмотрим на 100 самых частотных слов русского языка, собранные на основе данных НКРЯ, и проверим, какие из них отсутствуют в стоп-словах из библиотеки NLTK:

In [168]:
ruscorpora_freqs = ['и', 'в', 'не', 'на', 'я', 'быть', 'он', 'с', 'что', 'а', 'по', 'это', 'она', 'этот', 'к', 'но', 'они', 'мы', 'как', 'из', 'у', 'который', 'то', 'за', 'свой', 'что', 'весь', 'год', 'от', 'так', 'о', 'для', 'ты', 'же', 'все', 'тот', 'мочь', 'вы', 'человек', 'такой', 'его', 'сказать', 'только', 'или', 'ещё', 'бы', 'себя', 'один', 'как', 'уже', 'до', 'время', 'если', 'сам', 'когда', 'другой', 'вот', 'говорить', 'наш', 'мой', 'знать', 'стать', 'при', 'чтобы', 'дело', 'жизнь', 'кто', 'первый', 'очень', 'два', 'день', 'её', 'новый', 'рука', 'даже', 'во', 'со', 'раз', 'где', 'там', 'под', 'можно', 'ну', 'какой', 'после', 'их', 'работа', 'без', 'самый', 'потом', 'надо', 'хотеть', 'ли', 'слово', 'идти', 'большой', 'должен', 'место', 'иметь', 'ничто']

In [171]:
print([word for word in ruscorpora_freqs if word.replace('ё', 'е') not in stops])

['это', 'который', 'свой', 'весь', 'год', 'мочь', 'человек', 'сказать', 'время', 'говорить', 'наш', 'знать', 'стать', 'дело', 'жизнь', 'первый', 'очень', 'день', 'новый', 'рука', 'работа', 'самый', 'хотеть', 'слово', 'идти', 'большой', 'должен', 'место', 'иметь', 'ничто']


Я бы добавила следующие слова, потому что они относятся к служебным частям речи, обладают высокой частотностью и потому невысокой информативностью и формы некоторых из них есть в списке NLTK, но почему-то нет их самых:

1. _это_  
2. _который_  
3. _свой_  
4. _весь_  
5. _наш_

В целом список из NLTK нормальный, но странно было обнаружить, что как минимум пять перечисленных выше слов в нем отсутствуют и что некоторые слова есть не во всех формах, причем иногда слова нет в начальной, но есть в косвенных. Смущает, что совсем нет числительных и почти нет вводных слов. Лично я предпочитаю использовать список стоп-слов из библиотеки stop_words.

### Задание 5

In [221]:
from stop_words import get_stop_words
from pymystem3 import Mystem
import pymorphy2

In [222]:
stop_words = set(get_stop_words('ru'))

In [223]:
# лемматизируйте токены с помощью pymorphy2

In [224]:
morph = pymorphy2.MorphAnalyzer()

In [225]:
def get_pymorphy2_lemmatized_text(tokens):
    lemmatized_text = [morph.parse(token)[0].normal_form for token in tokens]
    lemmatized_text_without_stops = [lemma for lemma in lemmatized_text 
                                     if re.search('[а-яёa-z]', lemma) and lemma not in stop_words]
    return lemmatized_text_without_stops

In [226]:
pymorphy2_lemmatized_text = get_pymorphy2_lemmatized_text(tokens)

In [227]:
# лемматизируйте текст с помощью mystem3 

In [228]:
m = Mystem()

In [229]:
def get_mystem_lemmatized_text(text):
    lemmatized_text = [lemma for lemma in m.lemmatize(text)]
    lemmatized_text_without_stops = [lemma for lemma in lemmatized_text 
                                     if re.search('[а-яёa-z]', lemma) and lemma not in stop_words]
    return lemmatized_text_without_stops

In [230]:
mystem_lemmatized_text = get_mystem_lemmatized_text(' '.join(tokens))

In [231]:
sorted_by_freq(pymorphy2_lemmatized_text)

[('свой', 559),
 ('юрий', 426),
 ('андрей', 418),
 ('доктор', 383),
 ('стать', 332),
 ('говорить', 328),
 ('дом', 306),
 ('знать', 289),
 ('рука', 287),
 ('большой', 243),
 ('место', 223),
 ('дело', 218),
 ('друг', 213),
 ('лара', 213),
 ('самый', 211),
 ('думать', 199),
 ('сторона', 190),
 ('комната', 190),
 ('идти', 182),
 ('слово', 180),
 ('дорога', 180),
 ('видеть', 179),
 ('земля', 179),
 ('окно', 178),
 ('конец', 173),
 ('глаз', 170),
 ('ничто', 169),
 ('город', 163),
 ('нея', 156),
 ('что-то', 147),
 ('какой-то', 142),
 ('александр', 138),
 ('свет', 137),
 ('улица', 137),
 ('нога', 136),
 ('голова', 134),
 ('минута', 133),
 ('хороший', 130),
 ('поезд', 130),
 ('снег', 128),
 ('лицо', 128),
 ('сила', 126),
 ('часть', 125),
 ('ночь', 118),
 ('стол', 118),
 ('дать', 117),
 ('николай', 116),
 ('юра', 113),
 ('дверь', 112),
 ('вид', 111),
 ('двор', 110),
 ('голос', 109),
 ('должный', 108),
 ('мысль', 107),
 ('живаго', 103),
 ('чёрный', 102),
 ('москва', 102),
 ('новый', 101),
 ('посл

In [232]:
sorted_by_freq(mystem_lemmatized_text)

[('свой', 559),
 ('юрий', 426),
 ('андреевич', 418),
 ('доктор', 384),
 ('говорить', 375),
 ('становиться', 362),
 ('знать', 358),
 ('рука', 290),
 ('дом', 272),
 ('место', 225),
 ('стоять', 224),
 ('дело', 218),
 ('думать', 200),
 ('друг', 196),
 ('сторона', 194),
 ('комната', 190),
 ('лара', 185),
 ('голова', 182),
 ('идти', 182),
 ('слово', 181),
 ('видеть', 180),
 ('земля', 180),
 ('дорога', 180),
 ('самый', 179),
 ('окно', 178),
 ('конец', 173),
 ('глаз', 170),
 ('новый', 170),
 ('ничто', 167),
 ('город', 165),
 ('оставаться', 163),
 ('понимать', 159),
 ('ночь', 159),
 ('что-то', 156),
 ('давать', 156),
 ('свет', 144),
 ('часть', 143),
 ('какой-то', 142),
 ('сила', 141),
 ('лес', 141),
 ('большой', 140),
 ('улица', 138),
 ('нога', 138),
 ('выходить', 137),
 ('юра', 136),
 ('минута', 133),
 ('поезд', 133),
 ('снег', 128),
 ('лицо', 128),
 ('казаться', 125),
 ('пойти', 123),
 ('стол', 118),
 ('белый', 116),
 ('вид', 112),
 ('дверь', 112),
 ('проходить', 110),
 ('двор', 110),
 ('вода